src/pipdeptree/_models/package.py

```diff
@@ -118,31 +110,80 @@ class DistPackage(Package):
 
     """
 
-    def __init__(self, obj: DistInfoDistribution, req: ReqPackage | None = None) -> None:
-        super().__init__(obj)
+    def __init__(self, obj: Distribution, req: ReqPackage | None = None) -> None:
+        if hasattr(obj, "key"):
+            super().__init__(obj.key)
+        else:
+            super().__init__(obj.metadata["Name"])
```

Comment:
Looking at the most recent CI failures, I do see this test:
```console
___________________ test_dist_package_project_name_recovered ___________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fb15dd535c0>

    def test_dist_package_project_name_recovered(monkeypatch: pytest.MonkeyPatch) -> None:
        monkeypatch.setattr("pipdeptree._models.package.metadata", lambda _: {"name": "good"})
        foo = Mock(key="foo", project_name="bad", version="20.4.1")
        dp = DistPackage(foo)
>       assert dp.project_name == "good"
E       AssertionError: assert 'foo' == 'good'
E         
E         - good
E         + foo
```
This is failing because the mock object is given a `key` field, and with your code we first check if we have a key attribute and if we do we return it.

Like you said, `Distribution` and `Requirement` don't have a `key` field. So let's just do this:
```suggestion
    def __init__(self, obj: Distribution, req: ReqPackage | None = None) -> None:
        super().__init__(obj.metadata["Name"])
```

Reply:
I tried to remove the `key` field and it turns out to be another big work. So I suggest we let this part to be done in another PR? Including this change would make this PR more complicated.

Reply:
I made this change in `__init__` for the testcases compatibility entirely.

Reply:
Don't worry, I'm almost done with this part. Now I am solving the failed testcases.